pkg/agent/nplagent/rules/iptable_rule.go

```diff
@@ -0,0 +1,190 @@
+
+type IPTableRules struct {
+	name  string
+	table *iptables.IPTables
+}
+
+func NewIPTableRules() *IPTableRules {
+	iptables.New()
+	ipt, err := iptables.New()
+	if err != nil {
+		klog.Infof("init iptable for NPL failed: %v\n", err)
+		return nil
+	}
+	iptRule := IPTableRules{
+		name:  "NPL",
+		table: ipt,
+	}
+	return &iptRule
+}
+
+var once sync.Once
+
+func (ipt *IPTableRules) Init() bool {
+	if !ipt.DeleteAllRules() {
+		return false
+	}
+	return ipt.CreateChains()
+}
+
+// CreateChains : Create the chain NODE-PORT-LOCAL in NAT table
+// All DNAT rules for NPL would be added in this chain
+func (ipt *IPTableRules) CreateChains() bool {
+	exists, err := ipt.table.Exists("nat", NodePortLocalChain)
+	if err != nil {
+		klog.Warningf("check for NODE-PORT-LOCAL chain in iptable failed with error: %v", err)
+		return false
+	}
+	if !exists {
+		err = ipt.table.NewChain("nat", NodePortLocalChain)
+		if err != nil {
+			klog.Warningf("IPtable chain creation failed for NPL with error: %v", err)
+			return false
+		}
+	}
+
+	exists, err = ipt.table.Exists("nat", "PREROUTING", "-p", "tcp", "-j", NodePortLocalChain)
+	if err != nil {
+		klog.Warningf("check for NODE-PORT-LOCAL chain in iptable failed with error: %v", err)
+		return false
+	}
+	if !exists {
+		err = ipt.table.Append("nat", "PREROUTING", "-p", "tcp", "-j", NodePortLocalChain)
+		if err != nil {
+			klog.Warningf("IPtable rule creation in PREROUTING chain failed for NPL with error: %v", err)
+			return false
+		}
+	}
+	return true
+}
+
+// AddRule : Appends a DNAT rule in NODE-PORT-LOCAL chain of NAT table
+func (ipt *IPTableRules) AddRule(port int, podip string) bool {
+	exists, err := ipt.table.Exists("nat", NodePortLocalChain, "-p", "tcp", "-m", "tcp", "--dport",
+		fmt.Sprint(port), "-j", "DNAT", "--to-destination", podip)
+	if err != nil {
+		klog.Warningf("check for NODE-PORT-LOCAL chain in iptable failed with error: %v", err)
+		return false
+	}
+	if !exists {
+		err := ipt.table.Append("nat", NodePortLocalChain, "-p", "tcp", "-m", "tcp", "--dport",
+			fmt.Sprint(port), "-j", "DNAT", "--to-destination", podip)
+
+		if err != nil {
+			klog.Warningf("IPtable rule creation in failed for NPL with error: %v", err)
+			return false
+		}
+	}
+
+	return true
+}
+
+// DeleteRule : Delete a specific NPL rule from NODE-PORT-LOCAL chain
+func (ipt *IPTableRules) DeleteRule(port int, podip string) bool {
+	klog.Infof("Deleting rule with port %v and podip %v", port, podip)
+	err := ipt.table.Delete("nat", NodePortLocalChain, "-p", "tcp", "-m", "tcp", "--dport",
+		fmt.Sprint(port), "-j", "DNAT", "--to-destination", podip)
+
+	if err != nil {
+		klog.Infof("Failed to delete iptable rule for NPL: %v", err)
+		return false
+	}
+	return true
+}
+
+// SyncState : To Do - Compare existing rules with expected rules for all pods
+// and make sure that correct rules are programmed correctly
+func (ipt *IPTableRules) SyncState(podPort map[int]string) bool {
```

Comment:
I do not see this func is called in this PR?

Reply:
We are not using it now, that's why I had marked it as To-Do. But I can remove this.